test/vfd.c

```diff
@@ -3915,33 +4032,1949 @@ test_ctl(void)
+                                   void *write_bufs[], void *read_bufs[], char base_fill_char)
+{
+    hbool_t    result    = TRUE; /* will set to FALSE on failure */
+    char       fill_char = base_fill_char;
+    uint32_t   fix_point;
+    uint32_t   i;
+    uint32_t   j;
+    uint32_t   k;
+    H5FD_mem_t mem_types[6] = {H5FD_MEM_SUPER, H5FD_MEM_BTREE, H5FD_MEM_DRAW,
+                               H5FD_MEM_GHEAP, H5FD_MEM_LHEAP, H5FD_MEM_OHDR};
+
+    /* set the arrays of pointers to the write and read buffers to NULL,
+     * so that we can release memory on failure.
+     *
+     * Set the types[] and sizes[] arrays to invalid / improbable values
+     * so that use of these values will trigger failures.
      */
-    env_h5_drvr = HDgetenv(HDF5_DRIVER);
-    if (env_h5_drvr) {
-        HDprintf(" -- SKIPPED VFD tests because %s is set -- \n", HDF5_DRIVER);
-        HDexit(EXIT_SUCCESS);
+    for (i = 0; i < count; i++) {
+
+        write_bufs[i] = NULL;
+        read_bufs[i]  = NULL;
+        types[i]      = H5FD_MEM_NTYPES;
+        sizes[i]      = ULONG_MAX;
     }
 
-    h5_reset();
+    /* randomly select the point in the vector after which all entries are
+     * fixed at the same size and type.  Observe that 0 <= fix_point <
+     * count / 2.
+     */
+    fix_point = ((uint32_t)rand() & (count - 1)) / 2;
 
-    HDprintf("Testing basic Virtual File Driver functionality.\n");
+    HDassert(fix_point < count / 2);
+
+    for (i = 0; i < count; i++) {
+
+        if (i <= fix_point) {
+
+            types[i] = mem_types[i % 6];
+
+            addrs[i] = HADDR_UNDEF;
+
+            sizes[i] = (size_t)((rand() & 1023) + 1);
+
+            write_bufs[i] = HDmalloc(sizes[i] + 1);
+            read_bufs[i]  = HDmalloc(sizes[i] + 1);
+        }
+        else {
+
+            if (i == fix_point + 1) {
+
+                /* set the sentinels that indicate that all remaining
+                 * types and sizes are the same as the previous value.
+                 */
+                types[i] = H5FD_MEM_NOLIST;
+                sizes[i] = 0;
+            }
+
+            addrs[i] = HADDR_UNDEF;
+
+            write_bufs[i] = HDmalloc(sizes[fix_point] + 1);
+            read_bufs[i]  = HDmalloc(sizes[fix_point] + 1);
+        }
+
+        if ((NULL == write_bufs[i]) || (NULL == read_bufs[i])) {
+
+            HDfprintf(stderr, "%s: can't malloc read / write bufs.\n", __func__);
+            result = FALSE;
+            break;
+        }
+
+        /* need to avoid examining sizes beyond the fix_point */
+        k = MIN(i, fix_point);
+
+        for (j = 0; j < sizes[k]; j++) {
+
+            ((char *)(write_bufs[i]))[j] = fill_char;
+            ((char *)(read_bufs[i]))[j]  = '\0';
+        }
+
+        ((char *)(write_bufs[i]))[sizes[k]] = '\0';
+        ((char *)(read_bufs[i]))[sizes[k]]  = '\0';
+
+        fill_char++;
+    }
+
+    if (!result) { /* free buffers */
+
+        for (i = 0; i < count; i++) {
+
+            if (write_bufs[i]) {
```

Comment:
The standard says you can free NULL pointers, so the checks are just noise.

Reply:
I agree, though we've gone to great length to never free NULL elsewhere in the library. Is there an old system that couldn't handle this?